src/Orleans.Runtime/Core/LocalClient.cs

```diff
@@ -0,0 +1,218 @@
+                }
+                else
+                {
+                    extension = newExtensionFunc();
+                    var obj = this.grainFactory.CreateObjectReference<TExtensionInterface>(extension);
+
+                    addressable = obj;
+
+                    if (null == addressable)
+                    {
+                        throw new NullReferenceException("addressable");
+                    }
+                    entry = Tuple.Create((IGrainExtension) extension, addressable);
+                    this.extensionsTable.Add(typeof(TExtensionInterface), entry);
+                }
+            }
+
+            var typedAddressable = addressable.Cast<TExtensionInterface>();
+            // we have to return the extension as well as the IAddressable because the caller needs to root the extension
+            // to prevent it from being collected (the IAddressable uses a weak reference).
+            return Tuple.Create(extension, typedAddressable);
+        }
+
+        void IDisposable.Dispose()
+        {
+            if (this.disposing) return;
+            this.disposing = true;
+            Utils.SafeExecute(() => this.clientObserverRegistrar.ClientDropped(this.ClientId));
+            Utils.SafeExecute(() => this.clientObserverRegistrar.SetLocalClient(null));
+            Utils.SafeExecute(() => this.siloMessageCenter.SetLocalClient(null));
+            Utils.SafeExecute(() => this.listeningCts.Cancel(false));
+            Utils.SafeExecute(() => this.listeningCts.Dispose());
+        }
+        
+        public bool TryDispatchToClient(Message message)
+        {
+            if (!this.ClientId.Equals(message.TargetGrain)) return false;
+            if (message.IsExpired)
+            {
+                message.DropExpiredMessage(MessagingStatisticsGroup.Phase.Receive);
+                return true;
+            }
+
+            if (message.Direction == Message.Directions.Response)
+            {
+                // Requests are made through the runtime client, so deliver responses to the rutnime client so that the request callback can be executed.
+                this.runtimeClient.ReceiveResponse(message);
+            }
+            else
+            {
+                // Requests agrainst client objects are scheduled for execution on the client.
+                this.incomingMessages.Add(message);
+            }
+
+            return true;
+        }
+
+        private void Start()
+        {
+            var thread = new Thread(_ => this.RunClientMessagePump())
+            {
+                Name = nameof(LocalClient) + "." + nameof(RunClientMessagePump),
+                IsBackground = true,
+            };
+            thread.Start();
+        }
+
+        private void RunClientMessagePump()
+        {
+            while (!this.listeningCts.IsCancellationRequested)
+            {
+                try
+                {
+                    var message = this.incomingMessages.Take(this.listeningCts.Token);
```

Comment:
Note: this is not the most efficient implementation. To improve efficiency, we could share the existing ExecutorService with OrleansTaskScheduler. This is only called for invokable objects, though (regular responses are handled by InsideRuntimeClient), so it's not a high priority to optimize this just yet.